conveyor/conveyor.go

```diff
@@ -201,25 +189,80 @@ func (c *PulseConveyor) PreparePulse(pulse core.Pulse) error {
 	return nil
 }
 
+// PulseWithCallback contains info about new pulse and callback func
+type PulseWithCallback interface {
+	GetCallback() queue.SyncDone
+	GetPulse() core.Pulse
+	Done()
+}
+
+type pulseWithCallback struct {
+	callback queue.SyncDone
+	pulse    core.Pulse
+}
+
+// PulseWithCallback creates new instance of pulseWithCallback
+func NewPulseWithCallback(callback queue.SyncDone, pulse core.Pulse) PulseWithCallback {
+	return &pulseWithCallback{
+		callback: callback,
+		pulse:    pulse,
+	}
+}
+
+// GetCallback returns callback
+func (p *pulseWithCallback) GetCallback() queue.SyncDone {
+	return p.callback
+}
+
+// GetCallback returns callback
+func (p *pulseWithCallback) GetPulse() core.Pulse {
+	return p.pulse
+}
+
+// Done calls .Done() on func in callback param
+func (p *pulseWithCallback) Done() {
+	p.callback.Done()
+}
+
 // ActivatePulse activates conveyor with prepared pulse
 func (c *PulseConveyor) ActivatePulse() error {
 	if !c.IsOperational() {
 		return errors.New("[ ActivatePulse ] conveyor is not operational now")
 	}
 
 	c.lock.Lock()
-	defer c.lock.Unlock()
 
 	if c.futurePulseData == nil {
 		return errors.New("[ ActivatePulse ] preparation missing")
 	}
 
-	c.futurePulseData = nil
-
 	c.presentPulseNumber = c.futurePulseNumber
 	c.futurePulseNumber = c.newFuturePulseNumber
-	// TODO: add sending signal to slots queues
+
+	wg := sync.WaitGroup{}
+	wg.Add(2)
+
+	futureSlot := c.slotMap[*c.futurePulseNumber]
+	callback := NewPulseWithCallback(&wg, *c.futurePulseData)
+	err := futureSlot.inputQueue.PushSignal(ActivatePulseSignal, callback)
+	if err != nil {
+		return errors.Wrapf(err, "[ ActivatePulse ] can't send signal to future slot (for pulse %d)", c.futurePulseNumber)
+	}
+
+	presentSlot := c.slotMap[*c.presentPulseNumber]
+	err = presentSlot.inputQueue.PushSignal(ActivatePulseSignal, &wg)
+	if err != nil {
+		return errors.Wrapf(err, "[ ActivatePulse ] can't send signal to present slot (for pulse %d)", c.presentPulseNumber)
+	}
+
+	c.futurePulseData = nil
 	c.state = Active
+	c.lock.Unlock()
+	wg.Wait()
```

Comment:
are you sure about order?

Reply:
I mean c.lock.Unlock() and wg.Wait()

Reply:
I am

Reply:
We unlock conveyor, and can add tasks to slots queues in other streams, but slot itself decides what to do with input queue (like dont read it intil signal is processed)